ravenframework/SupervisedLearning/FeatureSelection/RFE.py

```diff
@@ -0,0 +1,815 @@
+        else:
+          support_ = supportParallel_
+
+    if nGroups > 1:
+      support_[:] = False
+      featuresForRanking = copy.deepcopy(groupFeaturesForRanking)
+      ranking_ = copy.deepcopy(groupRanking_)
+      supportOfSupport_ = copy.deepcopy(groupSupportOfSupport_)
+      for g in range(nGroups):
+        subGroupMask = np.where(supportCandidates[g] == True)
+        support_[subGroupMask] = True
+      self.raiseAMessage("After subGroupping stragety, number of candidate features are {}".format(np.sum(support_)))
+      # apply cross correlation if activated
+      if self.applyCrossCorrelation:
+        supportIndex = 0
+        for idx in range(len(supportOfSupport_)):
+          if mask[idx]:
+            supportOfSupport_[idx] = support_[supportIndex]
+            supportIndex=supportIndex+1
+        if self.whichSpace == 'feature':
+          features = np.arange(nFeatures)[supportOfSupport_]
+          targets = np.arange(nTargets)
+        else:
+          features = np.arange(nFeatures)
+          targets = np.arange(nTargets)[supportOfSupport_]
+
+        if self.whichSpace == 'feature':
+          space = X[:, features] if len(X.shape) < 3 else np.average(X[:, :,features],axis=0)
+        else:
+          space = y[:, targets] if len(y.shape) < 3 else  np.average(y[:, :,targets],axis=0)
+
+        from sklearn.feature_selection import mutual_info_regression
+        ti = None
+        ranksss = []
+        for g in range(nGroups):
+          tspace = y[:, outputSpaceToKeep[g]] if len(y.shape) < 3 else  np.average(y[:, :,outputSpaceToKeep[g]],axis=0)
+          mi = np.ravel(mutual_info_regression(space,tspace[:,0]))
+          if ti is None:
+            ti = mi
+          else:
+            ti = ti - mi
+          ranksss.append(np.ravel(np.argsort(-1*mi)))
+          print(mi[ranksss[-1]])
+
+        # compute spearman
+        # we fill nan with 1.0 (so the distance for such variables == 0 (will be discarded)
+        corr = np.nan_to_num(spearmanr(space,axis=0).correlation,nan=1.0)
+        corr = (corr + corr.T) / 2.
+        np.fill_diagonal(corr, 1)
+        # We convert the correlation matrix to a distance matrix before performing
+        # hierarchical clustering using Ward's linkage.
+
+        distance_matrix = 1. - np.abs(corr)
+        dist_linkage = hierarchy.ward(squareform(distance_matrix))
+
+        t = float('{:.3e}'.format(0.000001*np.max(dist_linkage)))
+
+        self.raiseAMessage("Applying hierarchical clustering on feature to eliminate possible collinearities")
+        self.raiseAMessage(f"Applying distance clustering tollerance of <{t}>")
+        cluster_ids = hierarchy.fcluster(dist_linkage, t, criterion="distance")
+        cluster_id_to_feature_ids = defaultdict(list)
+        for idx, cluster_id in enumerate(cluster_ids):
+          cluster_id_to_feature_ids[cluster_id].append(idx)
+        selected_features = [v[0] for v in cluster_id_to_feature_ids.values()]
```

Comment:
camelBack for variables " cluster_ids, cluster_id_to_feature_ids, selected_features"

Reply:
done